codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -0,0 +1,273 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http2;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.UnsupportedMessageTypeException;
+import io.netty.handler.codec.http.HttpServerUpgradeHandler.UpgradeEvent;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.internal.UnstableApi;
+
+import static io.netty.handler.logging.LogLevel.INFO;
+
+/**
+ * An HTTP/2 handler that maps HTTP/2 frames to {@link Http2Frame} objects and vice versa. For every incoming HTTP/2
+ * frame a {@link Http2Frame} object is created and propagated via {@link #channelRead}. Outgoing {@link Http2Frame}
+ * objects received via {@link #write} are converted to the HTTP/2 wire format.
+ *
+ * <p>A change in stream state is propagated through the channel pipeline as a user event via
+ * {@link Http2StreamStateEvent} objects. When a HTTP/2 stream first becomes active a {@link Http2StreamActiveEvent}
+ * and when it gets closed a {@link Http2StreamClosedEvent} is emitted.
+ *
+ * <p>Server-side HTTP to HTTP/2 upgrade is supported in conjunction with {@link Http2ServerUpgradeCodec}; the necessary
+ * HTTP-to-HTTP/2 conversion is performed automatically.
+ *
+ * <p><em>This API is very immature.</em> The Http2Connection-based API is currently preferred over
+ * this API. This API is targeted to eventually replace or reduce the need for the Http2Connection-based API.
+ */
+@UnstableApi
+public class Http2FrameCodec extends ChannelDuplexHandler {
+
+    private static final Http2FrameLogger HTTP2_FRAME_LOGGER = new Http2FrameLogger(INFO, Http2FrameCodec.class);
+
+    private final Http2ConnectionHandler http2Handler;
+
+    private ChannelHandlerContext ctx;
+    private ChannelHandlerContext http2HandlerCtx;
+
+    /**
+     * Construct a new handler.
+     *
+     * @param server {@code true} this is a server
+     */
+    public Http2FrameCodec(boolean server) {
+        this(server, new DefaultHttp2FrameWriter());
+    }
+
+    // Visible for testing
+    Http2FrameCodec(boolean server, Http2FrameWriter frameWriter) {
+        Http2Connection connection = new DefaultHttp2Connection(server);
+        frameWriter = new Http2OutboundFrameLogger(frameWriter, HTTP2_FRAME_LOGGER);
+        Http2ConnectionEncoder encoder = new DefaultHttp2ConnectionEncoder(connection, frameWriter);
+        Http2FrameReader reader = new Http2InboundFrameLogger(new DefaultHttp2FrameReader(), HTTP2_FRAME_LOGGER);
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Close this "DefaultHttp2FrameReader". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2095)